File: src/core/Shaders/emissive_sampling.fs
// =============================================================================
// EMISSIVE TRIANGLE SAMPLING
// =============================================================================
// This file contains direct lighting from emissive triangles (next-event estimation)
// Samples emissive geometry to reduce noise in scenes with emissive surfaces

// -----------------------------------------------------------------------------
// Emissive Triangle Sampling Structures
// -----------------------------------------------------------------------------

uniform sampler2D emissiveTriangleTexture;
uniform ivec2 emissiveTriangleTexSize;
uniform int emissiveTriangleCount;
uniform float emissiveBoost;

struct EmissiveSample {
	vec3 position;       // Sample point on emissive triangle
	vec3 normal;         // Normal at sample point
	vec3 emission;       // Emissive color * intensity
	vec3 direction;      // Direction from hit point to sample
	float distance;      // Distance to emissive surface
	float pdf;           // Probability density
	float area;          // Triangle area
	float cosThetaLight; // Cosine of angle on light surface
	bool valid;          // Whether sample is valid
};

// -----------------------------------------------------------------------------
// Helper Functions
// -----------------------------------------------------------------------------

// Sample a point on a triangle using barycentric coordinates
vec3 sampleTriangle( vec3 v0, vec3 v1, vec3 v2, vec2 xi ) {
	float sqrtU = sqrt( xi.x );
	float u = 1.0 - sqrtU;
	float v = xi.y * sqrtU;
	float w = 1.0 - u - v;
	return u * v0 + v * v1 + w * v2;
}

// Calculate triangle area
float triangleArea( vec3 v0, vec3 v1, vec3 v2 ) {
	vec3 edge1 = v1 - v0;
	vec3 edge2 = v2 - v0;
	return length( cross( edge1, edge2 ) ) * 0.5;
}

// Interpolate triangle normals using barycentric coordinates
vec3 interpolateNormal( vec3 n0, vec3 n1, vec3 n2, vec2 xi ) {
	float sqrtU = sqrt( xi.x );
	float u = 1.0 - sqrtU;
	float v = xi.y * sqrtU;
	float w = 1.0 - u - v;
	return normalize( u * n0 + v * n1 + w * n2 );
}

// Check if a material is emissive
bool isEmissive( RayTracingMaterial material ) {
	return material.emissiveIntensity > 0.0 && length( material.emissive ) > 0.0;
}

// Calculate emissive power of a triangle
float calculateEmissivePower( RayTracingMaterial material, float area ) {
	if( ! isEmissive( material ) )
		return 0.0;

	// Power = emissive radiance * area
	float avgEmissive = ( material.emissive.r + material.emissive.g + material.emissive.b ) / 3.0;
	return avgEmissive * material.emissiveIntensity * area;
}

// -----------------------------------------------------------------------------
// Triangle Data Access
// -----------------------------------------------------------------------------

// Fetch triangle vertices from texture
void fetchTriangleData(
	int triangleIndex,
	out vec3 v0, out vec3 v1, out vec3 v2,
	out vec3 n0, out vec3 n1, out vec3 n2,
	out int materialIndex
) {
	// Each triangle takes 8 vec4s (32 floats)
	int baseOffset = triangleIndex * 8;

	// Positions
	vec4 pos0 = texelFetch( triangleTexture, ivec2( baseOffset % triangleTexSize.x, baseOffset / triangleTexSize.x ), 0 );
	vec4 pos1 = texelFetch( triangleTexture, ivec2( ( baseOffset + 1 ) % triangleTexSize.x, ( baseOffset + 1 ) / triangleTexSize.x ), 0 );
	vec4 pos2 = texelFetch( triangleTexture, ivec2( ( baseOffset + 2 ) % triangleTexSize.x, ( baseOffset + 2 ) / triangleTexSize.x ), 0 );

	v0 = pos0.xyz;
	v1 = pos1.xyz;
	v2 = pos2.xyz;

	// Normals
	vec4 norm0 = texelFetch( triangleTexture, ivec2( ( baseOffset + 3 ) % triangleTexSize.x, ( baseOffset + 3 ) / triangleTexSize.x ), 0 );
	vec4 norm1 = texelFetch( triangleTexture, ivec2( ( baseOffset + 4 ) % triangleTexSize.x, ( baseOffset + 4 ) / triangleTexSize.x ), 0 );
	vec4 norm2 = texelFetch( triangleTexture, ivec2( ( baseOffset + 5 ) % triangleTexSize.x, ( baseOffset + 5 ) / triangleTexSize.x ), 0 );

	n0 = norm0.xyz;
	n1 = norm1.xyz;
	n2 = norm2.xyz;

	// Material index (stored in last vec4)
	vec4 uvMat = texelFetch( triangleTexture, ivec2( ( baseOffset + 7 ) % triangleTexSize.x, ( baseOffset + 7 ) / triangleTexSize.x ), 0 );
	materialIndex = int( uvMat.z );
}

// -----------------------------------------------------------------------------
// Emissive Triangle Sampling (Simple Uniform Sampling)
// -----------------------------------------------------------------------------

// Sample from emissive triangle index (100% success rate!)
// Uses pre-built list of emissive triangles for direct sampling
EmissiveSample sampleEmissiveTriangle(
	vec3 hitPoint,
	vec3 surfaceNormal,
	int totalTriangleCount,
	inout uint rngState
) {
	EmissiveSample result;
	result.valid = false;
	result.pdf = 0.0;

	// Check if we have emissive triangles
	if( emissiveTriangleCount <= 0 )
		return result;

	// Select random emissive triangle from the index
	float randEmissive = RandomValue( rngState );
	int emissiveIndex = int( randEmissive * float( emissiveTriangleCount ) );
	emissiveIndex = clamp( emissiveIndex, 0, emissiveTriangleCount - 1 );

	// Fetch emissive triangle data from texture
	// Texture layout: R=triangleIndex, G=power, B=cdf, A=unused
	ivec2 texCoord = ivec2(
		emissiveIndex % emissiveTriangleTexSize.x,
		emissiveIndex / emissiveTriangleTexSize.x
	);
	vec4 emissiveData = texelFetch( emissiveTriangleTexture, texCoord, 0 );
	int triangleIndex = int( emissiveData.r );
	float emissivePower = emissiveData.g;

	// Fetch triangle geometry
	vec3 v0, v1, v2, n0, n1, n2;
	int matIndex;
	fetchTriangleData( triangleIndex, v0, v1, v2, n0, n1, n2, matIndex );

	// Get material
	RayTracingMaterial material = getMaterial( matIndex );

	// Sample point on triangle
	vec2 xi = vec2( RandomValue( rngState ), RandomValue( rngState ) );
	vec3 samplePos = sampleTriangle( v0, v1, v2, xi );
	vec3 sampleNormal = interpolateNormal( n0, n1, n2, xi );

	// Direction from surface to emissive triangle
	vec3 toEmissive = samplePos - hitPoint;
	float distSq = dot( toEmissive, toEmissive );
	float dist = sqrt( distSq );
	vec3 dir = toEmissive / dist;

	// Check if facing the surface
	float surfaceFacing = dot( dir, surfaceNormal );
	float emissiveFacing = dot( dir, -sampleNormal );

	if( surfaceFacing <= 0.0 || emissiveFacing <= 0.0 )
		return result;

	// Calculate triangle area for PDF
	float area = triangleArea( v0, v1, v2 );

	// PDF for area sampling: 1 / area (uniform over triangle surface)
	// Convert to solid angle: pdf_solid = pdf_area * (distance^2 / cos(theta_light))
	// For uniform sampling over emissive triangles: pdf_area = 1 / (emissiveTriangleCount * area)
	float pdfArea = 1.0 / ( float( emissiveTriangleCount ) * area );
	float pdfSolidAngle = pdfArea * distSq / emissiveFacing;

	// Build result
	result.position = samplePos;
	result.normal = sampleNormal;
	result.emission = material.emissive * material.emissiveIntensity;
	result.direction = dir;
	result.distance = dist;
	result.pdf = max( pdfSolidAngle, MIN_PDF );
	result.area = area;
	result.cosThetaLight = emissiveFacing;
	result.valid = true;

	return result;
}

// -----------------------------------------------------------------------------
// Emissive Triangle Direct Lighting Contribution
// -----------------------------------------------------------------------------

struct EmissiveContributionResult {
	vec3 contribution;
	bool hasEmissive;
	vec3 emissionOnly;  // For debug visualization
	float distance;     // Distance to emissive surface
};

EmissiveContributionResult calculateEmissiveTriangleContributionDebug(
	vec3 hitPoint,
	vec3 normal,
	vec3 viewDir,
	RayTracingMaterial material,
	int totalTriangleCount,
	int bounceIndex,
	inout uint rngState,
	inout ivec2 stats
) {
	EmissiveContributionResult result;
	result.contribution = vec3( 0.0 );
	result.hasEmissive = false;
	result.emissionOnly = vec3( 0.0 );
	result.distance = 0.0;
	// Skip for very rough diffuse surfaces on secondary bounces (low contribution)
	if( bounceIndex > 1 && material.roughness > 0.9 && material.metalness < 0.1 )
		return result;

	// Sample emissive triangle
	EmissiveSample emissiveSample = sampleEmissiveTriangle( hitPoint, normal, totalTriangleCount, rngState );

	if( ! emissiveSample.valid || emissiveSample.pdf <= 0.0 )
		return result;

	result.hasEmissive = true;
	result.emissionOnly = emissiveSample.emission;
	result.distance = emissiveSample.distance;

	// Check geometric validity
	float NoL = max( 0.0, dot( normal, emissiveSample.direction ) );
	if( NoL <= 0.0 )
		return result;

	// Calculate ray offset for shadow ray
	vec3 rayOffset = calculateRayOffset( hitPoint, normal, material );
	vec3 rayOrigin = hitPoint + rayOffset;

	// Trace shadow ray
	float shadowDist = emissiveSample.distance - 0.001;
	float visibility = traceShadowRay( rayOrigin, emissiveSample.direction, shadowDist, rngState, stats );

	if( visibility <= 0.0 )
		return result;

	// Evaluate BRDF
	vec3 brdfValue = evaluateMaterialResponse( viewDir, emissiveSample.direction, normal, material );

	// For emissive surfaces, we're sampling radiance (not intensity)
	// The contribution from an area light with radiance L is:
	// dL = L * BRDF * cos(θ_surface) * cos(θ_light) * (area / distance²) * visibility
	//
	// But our emissiveSample.emission is already the radiance, so:
	// contribution = emission * BRDF * cos(θ_surface) * cos(θ_light) * (area / distance²)

	float distSq = emissiveSample.distance * emissiveSample.distance;
	float solidAngleTerm = emissiveSample.area * emissiveSample.cosThetaLight / distSq;

	// Final contribution (no 4π because we're not converting from intensity)
	// Note: This bypasses firefly suppression in pathtracer_core.fs
	// Large boost needed because: BRDF (~0.1-0.3) * geometric falloff * cosine terms = very small
	// emissiveBoost is a uniform that can be adjusted via UI
	result.contribution = emissiveSample.emission * brdfValue * NoL * solidAngleTerm * visibility * emissiveBoost;

	return result;
}

// Wrapper function for backward compatibility
vec3 calculateEmissiveTriangleContribution(
	vec3 hitPoint,
	vec3 normal,
	vec3 viewDir,
	RayTracingMaterial material,
	int totalTriangleCount,
	int bounceIndex,
	inout uint rngState,
	inout ivec2 stats
) {
	EmissiveContributionResult result = calculateEmissiveTriangleContributionDebug(
		hitPoint, normal, viewDir, material, totalTriangleCount, bounceIndex, rngState, stats
	);
	return result.contribution;
}

File: src/core/WebGPU/TSL/EmissiveSampling.js
// Emissive Triangle Sampling - Ported from emissive_sampling.fs
// Direct lighting from emissive triangles (next-event estimation)

import {
	Fn,
	vec2,
	vec3,
	vec4,
	float,
	int,
	bool as tslBool,
	If,
	dot,
	normalize,
	cross,
	length,
	max,
	sqrt,
	clamp,
} from 'three/tsl';

import { struct } from './structProxy.js';
import { RayTracingMaterial, HitInfo } from './Struct.js';
import { MIN_PDF, getDatafromDataTexture, getMaterial } from './Common.js';
import { RandomValue } from './Random.js';

// ================================================================================
// STRUCTS
// ================================================================================

export const EmissiveSample = struct( {
	position: 'vec3', // Sample point on emissive triangle
	normal: 'vec3', // Normal at sample point
	emission: 'vec3', // Emissive color * intensity
	direction: 'vec3', // Direction from hit point to sample
	distance: 'float', // Distance to emissive surface
	pdf: 'float', // Probability density
	area: 'float', // Triangle area
	cosThetaLight: 'float', // Cosine of angle on light surface
	valid: 'bool', // Whether sample is valid
} );

export const EmissiveContributionResult = struct( {
	contribution: 'vec3',
	hasEmissive: 'bool',
	emissionOnly: 'vec3', // For debug visualization
	distance: 'float', // Distance to emissive surface
} );

// ================================================================================
// HELPER FUNCTIONS
// ================================================================================

// Sample a point on a triangle using barycentric coordinates
export const sampleTriangle = Fn( ( [ v0, v1, v2, xi ] ) => {

	const sqrtU = sqrt( xi.x );
	const u = float( 1.0 ).sub( sqrtU );
	const v = xi.y.mul( sqrtU );
	const w = float( 1.0 ).sub( u ).sub( v );
	return v0.mul( u ).add( v1.mul( v ) ).add( v2.mul( w ) );

} );

// Calculate triangle area
export const triangleArea = Fn( ( [ v0, v1, v2 ] ) => {

	const edge1 = v1.sub( v0 );
	const edge2 = v2.sub( v0 );
	return length( cross( edge1, edge2 ) ).mul( 0.5 );

} );

// Interpolate triangle normals using barycentric coordinates
export const interpolateNormal = Fn( ( [ n0, n1, n2, xi ] ) => {

	const sqrtU = sqrt( xi.x );
	const u = float( 1.0 ).sub( sqrtU );
	const v = xi.y.mul( sqrtU );
	const w = float( 1.0 ).sub( u ).sub( v );
	return normalize( n0.mul( u ).add( n1.mul( v ) ).add( n2.mul( w ) ) );

} );

// Check if a material is emissive
export const isEmissive = Fn( ( [ material ] ) => {

	return material.emissiveIntensity.greaterThan( 0.0 ).and( length( material.emissive ).greaterThan( 0.0 ) );

} );

// Calculate emissive power of a triangle
export const calculateEmissivePower = Fn( ( [ material, area ] ) => {

	const result = float( 0.0 ).toVar( 'emPower' );

	If( isEmissive( material ), () => {

		const avgEmissive = material.emissive.x.add( material.emissive.y ).add( material.emissive.z ).div( 3.0 );
		result.assign( avgEmissive.mul( material.emissiveIntensity ).mul( area ) );

	} );

	return result;

} );

// ================================================================================
// TRIANGLE DATA ACCESS
// ================================================================================

const TRI_STRIDE = 8;

const TriangleData = struct( {
	v0: 'vec3', v1: 'vec3', v2: 'vec3',
	n0: 'vec3', n1: 'vec3', n2: 'vec3',
	materialIndex: 'int',
} );

// Fetch triangle vertices from texture
// Returns data packed in a struct-like way
export const fetchTriangleData = Fn( ( [ triangleIndex, triangleTexture, triangleTexSize ] ) => {

	// Positions
	const pos0 = getDatafromDataTexture( triangleTexture, triangleTexSize, triangleIndex, int( 0 ), int( TRI_STRIDE ) );
	const pos1 = getDatafromDataTexture( triangleTexture, triangleTexSize, triangleIndex, int( 1 ), int( TRI_STRIDE ) );
	const pos2 = getDatafromDataTexture( triangleTexture, triangleTexSize, triangleIndex, int( 2 ), int( TRI_STRIDE ) );

	// Normals
	const norm0 = getDatafromDataTexture( triangleTexture, triangleTexSize, triangleIndex, int( 3 ), int( TRI_STRIDE ) );
	const norm1 = getDatafromDataTexture( triangleTexture, triangleTexSize, triangleIndex, int( 4 ), int( TRI_STRIDE ) );
	const norm2 = getDatafromDataTexture( triangleTexture, triangleTexSize, triangleIndex, int( 5 ), int( TRI_STRIDE ) );

	// Material index (stored in last vec4)
	const uvMat = getDatafromDataTexture( triangleTexture, triangleTexSize, triangleIndex, int( 7 ), int( TRI_STRIDE ) );

	// Return all data as a struct
	return TriangleData( {
		v0: pos0.xyz, v1: pos1.xyz, v2: pos2.xyz,
		n0: norm0.xyz, n1: norm1.xyz, n2: norm2.xyz,
		materialIndex: int( uvMat.z ),
	} );

} );

// ================================================================================
// EMISSIVE TRIANGLE SAMPLING
// ================================================================================

// Sample from emissive triangle index
export const sampleEmissiveTriangle = Fn( ( [
	hitPoint, surfaceNormal, totalTriangleCount,
	rngState,
	emissiveTriangleTexture, emissiveTriangleTexSize, emissiveTriangleCount,
	triangleTexture, triangleTexSize,
	materialTexture, materialTexSize,
] ) => {

	const result = EmissiveSample( {
		position: vec3( 0.0 ),
		normal: vec3( 0.0 ),
		emission: vec3( 0.0 ),
		direction: vec3( 0.0 ),
		distance: float( 0.0 ),
		pdf: float( 0.0 ),
		area: float( 0.0 ),
		cosThetaLight: float( 0.0 ),
		valid: false,
	} ).toVar( 'emissiveSample' );

	// Check if we have emissive triangles
	If( emissiveTriangleCount.greaterThan( int( 0 ) ), () => {

		// Select random emissive triangle from the index
		const randEmissive = RandomValue( rngState );
		const emissiveIndex = clamp(
			int( randEmissive.mul( float( emissiveTriangleCount ) ) ),
			int( 0 ),
			emissiveTriangleCount.sub( 1 )
		).toVar( 'emIdx' );

		// Fetch emissive triangle data from texture
		// Texture layout: R=triangleIndex, G=power, B=cdf, A=unused
		const emissiveTexCoord = vec2(
			float( emissiveIndex.modInt( emissiveTriangleTexSize.x ) ).add( 0.5 ).div( float( emissiveTriangleTexSize.x ) ),
			float( emissiveIndex.div( emissiveTriangleTexSize.x ) ).add( 0.5 ).div( float( emissiveTriangleTexSize.y ) )
		);
		const emissiveData = emissiveTriangleTexture.sample( emissiveTexCoord );
		const triangleIndex = int( emissiveData.r );

		// Fetch triangle geometry
		const triData = TriangleData.wrap( fetchTriangleData( triangleIndex, triangleTexture, triangleTexSize ) );

		// Get material
		const material = RayTracingMaterial.wrap( getMaterial( triData.materialIndex, materialTexture, materialTexSize ) );

		// Sample point on triangle
		const xi = vec2( RandomValue( rngState ), RandomValue( rngState ) );
		const samplePos = sampleTriangle( triData.v0, triData.v1, triData.v2, xi );
		const sampleNormal = interpolateNormal( triData.n0, triData.n1, triData.n2, xi );

		// Direction from surface to emissive triangle
		const toEmissive = samplePos.sub( hitPoint );
		const distSq = dot( toEmissive, toEmissive );
		const dist = sqrt( distSq );
		const dir = toEmissive.div( dist );

		// Check if facing the surface
		const surfaceFacing = dot( dir, surfaceNormal );
		const emissiveFacing = dot( dir, sampleNormal.negate() );

		If( surfaceFacing.greaterThan( 0.0 ).and( emissiveFacing.greaterThan( 0.0 ) ), () => {

			// Calculate triangle area for PDF
			const area = triangleArea( triData.v0, triData.v1, triData.v2 );

			// PDF for area sampling converted to solid angle
			const pdfArea = float( 1.0 ).div( float( emissiveTriangleCount ).mul( area ) );
			const pdfSolidAngle = pdfArea.mul( distSq ).div( emissiveFacing );

			// Build result
			result.position.assign( samplePos );
			result.normal.assign( sampleNormal );
			result.emission.assign( material.emissive.mul( material.emissiveIntensity ) );
			result.direction.assign( dir );
			result.distance.assign( dist );
			result.pdf.assign( max( pdfSolidAngle, MIN_PDF ) );
			result.area.assign( area );
			result.cosThetaLight.assign( emissiveFacing );
			result.valid.assign( true );

		} );

	} );

	return result;

} );

// ================================================================================
// EMISSIVE TRIANGLE DIRECT LIGHTING CONTRIBUTION
// ================================================================================

// Note: calculateEmissiveTriangleContributionDebug requires traceShadowRay and
// evaluateMaterialResponse which creates circular dependencies.
// These are passed as function parameters to avoid the cycle.

export const calculateEmissiveTriangleContributionDebug = Fn( ( [
	hitPoint, normal, viewDir, material,
	totalTriangleCount, bounceIndex, rngState,
	emissiveBoost,
	emissiveTriangleTexture, emissiveTriangleTexSize, emissiveTriangleCount,
	triangleTexture, triangleTexSize,
	materialTexture, materialTexSize,
	// Callback functions to avoid circular deps
	traceShadowRayFn,
	evaluateMaterialResponseFn,
	calculateRayOffsetFn,
] ) => {

	const result = EmissiveContributionResult( {
		contribution: vec3( 0.0 ),
		hasEmissive: false,
		emissionOnly: vec3( 0.0 ),
		distance: float( 0.0 ),
	} ).toVar( 'emContrib' );

	// Skip for very rough diffuse surfaces on secondary bounces
	const skip = bounceIndex.greaterThan( int( 1 ) ).and( material.roughness.greaterThan( 0.9 ) ).and( material.metalness.lessThan( 0.1 ) );

	If( skip.not(), () => {

		// Sample emissive triangle
		const emissiveSample = EmissiveSample.wrap( sampleEmissiveTriangle(
			hitPoint, normal, totalTriangleCount, rngState,
			emissiveTriangleTexture, emissiveTriangleTexSize, emissiveTriangleCount,
			triangleTexture, triangleTexSize,
			materialTexture, materialTexSize,
		) );

		If( emissiveSample.valid.and( emissiveSample.pdf.greaterThan( 0.0 ) ), () => {

			result.hasEmissive.assign( true );
			result.emissionOnly.assign( emissiveSample.emission );
			result.distance.assign( emissiveSample.distance );

			// Check geometric validity
			const NoL = max( float( 0.0 ), dot( normal, emissiveSample.direction ) );

			If( NoL.greaterThan( 0.0 ), () => {

				// Calculate ray offset for shadow ray
				const rayOffset = calculateRayOffsetFn( hitPoint, normal, material );
				const rayOrigin = hitPoint.add( rayOffset );

				// Trace shadow ray
				const shadowDist = emissiveSample.distance.sub( 0.001 );
				const visibility = traceShadowRayFn( rayOrigin, emissiveSample.direction, shadowDist, rngState );

				If( visibility.greaterThan( 0.0 ), () => {

					// Evaluate BRDF
					const brdfValue = evaluateMaterialResponseFn( viewDir, emissiveSample.direction, normal, material );

					// Calculate solid angle term
					const distSq = emissiveSample.distance.mul( emissiveSample.distance );
					const solidAngleTerm = emissiveSample.area.mul( emissiveSample.cosThetaLight ).div( distSq );

					// Final contribution
					result.contribution.assign(
						emissiveSample.emission.mul( brdfValue ).mul( NoL ).mul( solidAngleTerm ).mul( visibility ).mul( emissiveBoost )
					);

				} );

			} );

		} );

	} );

	return result;

} );

// Wrapper function for backward compatibility
export const calculateEmissiveTriangleContribution = Fn( ( [
	hitPoint, normal, viewDir, material,
	totalTriangleCount, bounceIndex, rngState,
	emissiveBoost,
	emissiveTriangleTexture, emissiveTriangleTexSize, emissiveTriangleCount,
	triangleTexture, triangleTexSize,
	materialTexture, materialTexSize,
	traceShadowRayFn,
	evaluateMaterialResponseFn,
	calculateRayOffsetFn,
] ) => {

	const result = EmissiveContributionResult.wrap( calculateEmissiveTriangleContributionDebug(
		hitPoint, normal, viewDir, material,
		totalTriangleCount, bounceIndex, rngState,
		emissiveBoost,
		emissiveTriangleTexture, emissiveTriangleTexSize, emissiveTriangleCount,
		triangleTexture, triangleTexSize,
		materialTexture, materialTexSize,
		traceShadowRayFn,
		evaluateMaterialResponseFn,
		calculateRayOffsetFn,
	) );
	return result.contribution;

} );
